hazelcast/src/main/java/com/hazelcast/config/YamlConfigBuilder.java

```diff
@@ -0,0 +1,302 @@
+     * instance is created and the config is located in every possible
+     * places. For these places, please see {@link YamlConfigLocator}.
+     * <p/>
+     * If the provided {@link YamlConfigLocator} is not {@code null}, it
+     * is expected that it already located the configuration YAML to load
+     * from. No further attempt to locate the configuration YAML is made
+     * if the configuration YAML is not located already.
+     *
+     * @param locator the configured locator to use
+     */
+    public YamlConfigBuilder(YamlConfigLocator locator) {
+        if (locator == null) {
+            locator = new YamlConfigLocator();
+            locator.locateFromSystemProperty();
+        }
+
+        this.in = locator.getIn();
+        this.configurationFile = locator.getConfigurationFile();
+        this.configurationUrl = locator.getConfigurationUrl();
+    }
+
+    @Override
+    public Config build() {
+        return build(new Config());
+    }
+
+    Config build(Config config) {
+        config.setConfigurationFile(configurationFile);
+        config.setConfigurationUrl(configurationUrl);
+        try {
+            parseAndBuildConfig(config);
+        } catch (Exception e) {
+            throw ExceptionUtil.rethrow(e);
+        }
+        return config;
+    }
+
+    private void parseAndBuildConfig(Config config) throws Exception {
+        YamlMapping yamlRootNode;
+        try {
+            yamlRootNode = ((YamlMapping) YamlLoader.load(in));
+        } catch (Exception ex) {
+            throw new InvalidConfigurationException("Invalid YAML configuration", ex);
+        }
+
+        YamlNode imdgRoot = yamlRootNode.childAsMapping(ConfigSections.HAZELCAST.name().toLowerCase());
+        if (imdgRoot == null) {
+            throw new InvalidConfigurationException("No mapping with hazelcast key is found in the provided configuration");
+        }
+
+        Node w3cRootNode = asW3cNode(imdgRoot);
+        replaceVariables(w3cRootNode);
+        importDocuments(imdgRoot);
+
+        new YamlMemberDomConfigProcessor(true, config).buildConfig(w3cRootNode);
+    }
+
+    private void importDocuments(YamlNode imdgRoot) throws Exception {
+        YamlMapping rootAsMapping = asMapping(imdgRoot);
+        YamlSequence importSeq = rootAsMapping.childAsSequence(ConfigSections.IMPORT.name);
+        if (importSeq == null || importSeq.childCount() == 0) {
+            return;
+        }
+
+        for (YamlNode importNode : importSeq.children()) {
+            String resource = asScalar(importNode).nodeValue();
+            URL url = ConfigLoader.locateConfig(resource);
+            if (url == null) {
+                throw new InvalidConfigurationException("Failed to load resource: " + resource);
+            }
+            if (!currentlyImportedFiles.add(url.getPath())) {
+                throw new InvalidConfigurationException("Cyclic loading of resource '" + url.getPath() + "' detected!");
```

Comment:
Probably it's better to rephrase the error message slightly, since the exception may also indicate duplicate imports, not necessarily cyclic.

Reply:
Rephrased in YAML and XML builders: 
```
throw new InvalidConfigurationException("Resource '" + url.getPath() + "' is already loaded! This can be due to"
+ " duplicate or cyclic imports.");
```